x/trigger/keeper/event_detector.go

```diff
@@ -0,0 +1,70 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/provenance-io/provenance/x/trigger/types"
+)
+
+// DetectBlockEvents Detects triggers that have been activated by their corresponding events.
+func (k Keeper) DetectBlockEvents(ctx sdk.Context) {
+	triggers := k.detectTransactionEvents(ctx)
+	triggers = append(triggers, k.detectBlockHeightEvents(ctx)...)
+	triggers = append(triggers, k.detectTimeEvents(ctx)...)
+
+	for _, trigger := range triggers {
+		k.UnregisterTrigger(ctx, trigger)
+		k.QueueTrigger(ctx, trigger)
+	}
+}
+
+// DetectTransactionEvents Detects triggers that have been activated by transaction events.
+func (k Keeper) detectTransactionEvents(ctx sdk.Context) (triggers []types.Trigger) {
+	detectedTriggers := map[uint64]bool{}
+	for _, event := range ctx.EventManager().GetABCIEventHistory() {
+		matched := k.getMatchingTriggers(ctx, event.GetType(), func(trigger types.Trigger, triggerEvent types.TriggerEventI) bool {
+			if _, isDetected := detectedTriggers[trigger.Id]; isDetected {
+				return false
+			}
+			txEvent := triggerEvent.(*types.TransactionEvent)
+			detected := txEvent.Equals(event)
+			detectedTriggers[trigger.Id] = detected
+			return detected
+		})
+		triggers = append(triggers, matched...)
+	}
```

Comment:
A block can easily have a significant number of events, on the order of 100s. We probably don't want to read each trigger from state and marshal it for each attribute.

It's probably best to put the event loop inside the `getMatchingTriggers` condition. Call `events := ctx.EventManager().GetABCIEventHistory()` first (before `getMatchingTriggers`) and loop over that each time.